tuxemon/mod_manager.py

```diff
@@ -0,0 +1,190 @@
+            return json.loads(file.read())
+
+    def cache_to_pkglist(self):
+        """
+        Override self.packages with the cache.
+        Uses self.packages = self.read_from_cache()
+        """
+        self.packages = self.read_from_cache()
+
+    def update(self, url):
+        """Returns the response from the server"""
+        with urllib.request.urlopen(url + "/api/packages") as packages:
+            return json.loads(packages.read().decode("UTF-8"))
+
+    def update_all(self):
+        """
+        Updates all packages in self.package.
+        It automatically clears the self.package variable, and then
+        populates it from the data from the repositories.
+        """
+        self.packages = []
+
+        for i in self.url:
+            for package in self.update(i):
+                self.packages.append(package)
+
+        self.write_to_cache()
+
+    def list_packages(self):
+        """Returns package dictionary, either from the server or the cache"""
+        return self.packages
+
+    def download_package(self, name, release, repo=None, dont_extract=False, install_deps=True):
+        """Downloads the specified package"""
+        if repo is None:
+            repo = self.get_package_repo(name)
+
+            # Remove trailing slash
+            if repo[-1] == "/":
+                repo = repo[:-1]
+
+        url = str(repo) + f"/packages/{name}/releases/{str(release)}/download"
+        filename = os.path.join(paths.CACHE_DIR + f"/downloaded_packages/{name}.{release}.zip")
+
+        # Apperantly this function is ported from urllib from python2.
+        # Maybe replace this in the future?
+        # https://docs.python.org/3/library/urllib.request.html#urllib.request.urlretrieve
+        urllib.request.urlretrieve(url, filename=filename)
+
+        outfolder = os.path.join(paths.BASEDIR, "mods", f"{name}")
+
+        self.write_package_to_list(outfolder, name)
+
+        if not dont_extract:
+            self.extract(filename, outfolder)
+            with open(f"{outfolder}/meta.json", "w") as metafile:
+                meta = self.get_package_info(name, repo)
+                metafile.write(json.dumps(meta, indent=4, sort_keys=False))
+
+        if install_deps:
+            # This function calls download_package, might cause issues
+            self.install_dependencies(name, release, repo, dont_extract=dont_extract)
+
+    def install_dependencies(self, name, release, repo, dont_extract=False, symlink=True):
+        """
+        Same as the download_package(), but it includes dependency installing.
+        When symlink is True, dependency's files will be linked.
+        """
+        # Get info
+        meta = self.get_package_info(name, repo)
+        if "dependencies" in meta:
+            dependencies = meta["dependencies"]
+
+            for pack in dependencies:
+                self.download_package(pack, release, repo, dont_extract=dont_extract, install_deps=False)
```

Comment:
doesn't look like it can detect cyclic redundancies?  a -> b -> c -> a would create a loop, it seems